stdlib/regex.gr

```diff
@@ -0,0 +1,2946 @@
+/* Grain Regular Expression Library
+ *
+ * This library provides support for regular expressions in Grain.
+ * Its parser and analyzer are largely ported from Racket (https://racket-lang.org/),
+ * which is licensed under Apache 2.0. Racket's regular expression
+ * engine is itself inspired by the Spencer engine, as found in Tcl.
+ */
+import Array from "array"
+import Char from "char"
+import List from "list"
+import Map from "map"
+import Option from "option"
+import Result from "result"
+import String from "string"
+import Float32 from "float32"
+
+/*
+
+===============================
+REGEX PARSER CONFIG DEFINITIONS
+===============================
+
+*/
+
+record RegExParserConfig {
+  isPerlRegExp: Bool,
+  caseSensitive: Bool,
+  multiLine: Bool,
+  groupNumber: Box<Number>,
+  references: Box<Bool>,
```

Comment:
It's actually important here to use boxes, due to the way that `withConfig` and friends work (since we create new records which share the underlying references). I can add a comment.